src/loam/memory.rs

```diff
@@ -0,0 +1,752 @@
+use std::cell::RefCell;
+
+use ascent::Dual;
+use itertools::Itertools;
+use p3_field::{AbstractField, PrimeField32};
+use rustc_hash::{FxHashMap, FxHashSet};
+
+use crate::{
+    loam::{allocation::Allocator, LEWrap, Num, Ptr, PtrEq, Wide, WidePtr, LE},
+    lurk::{
+        chipset::LurkChip,
+        eval::EvalErr,
+        state::{StateRcCell, LURK_PACKAGE_SYMBOLS_NAMES},
+        tag::Tag,
+        zstore::{self, builtin_vec, lurk_zstore, ZPtr, ZStore},
+    },
+};
+
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+pub struct Memory {
+    pub cons_digest_mem: Vec<(Wide, LE)>,
+    pub cons_mem: Vec<(Ptr, Ptr, LE)>,
+    pub fun_digest_mem: Vec<(Wide, LE)>,
+    pub fun_mem: Vec<(Ptr, Ptr, Ptr, LE)>,
+    pub thunk_digest_mem: Vec<(Wide, LE)>,
+    pub thunk_mem: Vec<(Ptr, Ptr, LE)>,
+
+    pub sym_digest_mem: Vec<(Wide, LE)>,
+    pub builtin_digest_mem: Vec<(Wide, LE)>,
+    pub nil_digest_mem: Vec<(Wide, LE)>,
+}
+
+impl Memory {
+    fn report_sizes(&self, summary: &mut DistillationSummary) {
+        summary.set_distilled_size(Tag::Cons, self.cons_mem.len());
+        summary.set_distilled_size(Tag::Fun, self.fun_mem.len());
+        summary.set_distilled_size(Tag::Thunk, self.thunk_mem.len());
+    }
+}
+
+#[repr(transparent)]
+#[derive(Clone, Copy, Debug, Ord, PartialOrd, PartialEq, Eq, Hash)]
+/// Virtual pointer
+pub struct VPtr(pub Ptr);
+
+impl VPtr {
+    fn new(tag: Tag, addr: u32) -> Self {
+        VPtr(Ptr(tag.elt(), LE::from_canonical_u32(addr)))
+    }
+
+    fn num(addr: u32) -> Self {
+        VPtr::new(Tag::Num, addr)
+    }
+
+    fn cons(addr: u32) -> Self {
+        VPtr::new(Tag::Cons, addr)
+    }
+
+    fn fun(addr: u32) -> Self {
+        VPtr::new(Tag::Fun, addr)
+    }
+
+    fn to_pptr(&self) -> PPtr {
+        PPtr(self.0)
+    }
+
+    fn tag(&self) -> Tag {
+        self.0.tag()
+    }
+
+    fn addr(&self) -> LE {
+        self.0 .1
+    }
+}
+
+#[repr(transparent)]
+#[derive(Clone, Copy, Debug, Ord, PartialOrd, PartialEq, Eq, Hash)]
+/// Physical pointer
+pub struct PPtr(pub Ptr);
+
+impl PPtr {
+    fn new(tag: Tag, addr: u32) -> Self {
+        PPtr(Ptr(tag.elt(), LE::from_canonical_u32(addr)))
+    }
+
+    fn num(addr: u32) -> Self {
+        PPtr::new(Tag::Num, addr)
+    }
+
+    fn cons(addr: u32) -> Self {
+        PPtr::new(Tag::Cons, addr)
+    }
+
+    fn fun(addr: u32) -> Self {
+        PPtr::new(Tag::Fun, addr)
+    }
+
+    fn tag(&self) -> Tag {
+        self.0.tag()
+    }
+
+    fn addr(&self) -> LE {
+        self.0 .1
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub enum PPtrType {
+    Tuple2(PPtr, PPtr),
+    Tuple3(PPtr, PPtr, PPtr),
+}
```

Comment:
This struct is confusingly named. Can you come up with something that doesn't make usage appear to be naming a *type*?

Reply:
I'll rename to `PPtrKind`. For more context, it's following the convention with `ZPtr`/`ZPtrType`, [here](https://github.com/argumentcomputer/lurk/blob/bfe044b80854b51e128317962a4b2ec52eb555fa/src/lurk/zstore.rs#L228-L238).